sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -1413,6 +1413,14 @@ object SQLConf {
       .booleanConf
       .createWithDefault(true)
 
+  val CONTINUOUS_STREAMING_EPOCH_BACKLOG_QUEUE_SIZE =
+    buildConf("spark.sql.streaming.continuous.epochBacklogQueueSize")
+      .internal()
+      .doc("The max number of entries to be stored in queue to wait for late epochs. " +
+        "If this parameter is exceeded by the size of the queue, stream will stop with an error.")
+      .intConf
+      .createWithDefault(10000)
```

Comment:
Current behavior is to not have a limit, so wouldn't it be better to have this be `Int.MaxValue`?

Reply:
That's the main purpose of this PR to add this limit. If the user is having a backlog > 10k then the query will never come back to live. Without this change the user can guess what is happening but the cluster is not progressing. With this change the query will throw an exception and the query can be restarted (maybe an intermittent issue). So this change is
* To tell the user what happened
* Fail fast(er)

Reply:
That's an issue because you're assuming that 10k is the number that will cause a problem - i.e. nobody before this change was able to properly run a query if this queue exceeded 10k items.

Meaning, that if this is checked in, all queries that feel into that case (more than 10k items but not running into issues) will now break.

This also goes together with the "why is the config internal" question. What is the reasoning for forcing this 10k value onto people, and not providing a clear way for them to override it?

Reply:
Yeah, I agree these 2 things together is bad. Even if 10k items on test clusters end up in really bad shaped nodes it doesn't mean the same scenario applies to all other clusters. The configuration should be provided for users for sure and I'll add it in a commit.

Having a pre-defined number I still think it has a value to notify users about a possibly faulty query. Having `Int.MaxValue` and queuing up things may end-up in silently stopped query but throwing and exception will end-up in a failing query where resource managers can restart it. If you still have concerns maybe the number can be adjusted.

Reply:
> Having a pre-defined number I still think it has a value to notify users about a possibly faulty query

That still doesn't explain why is it OK to potentially break things that were working fine.

Do you have statistics that show that this queue almost never gets close to this threshold, unless the application is for some reason unhealthy?

I understand that the previous behavior when things fail is bad. But it's worse to just break stuff that is not broken.

Reply:
e.g. you could leave it at `Int.MaxValue` to avoid breaking existing apps, but add a warning that this configuration can lead to OOM errors that are hard to debug.

Reply:
If the cluster is healthy and all partitions are processing well then the queue is not building up. The problem comes when for example one partition stuck somehow. The worst case if this partition is dead and 10k items from all other partitions are processed and waiting for this single one. Jose stated the same thing [here](https://github.com/apache/spark/pull/21392#discussion_r190112873). Statistics only from local cluster which is definitely not PROD stuff.

Partition stuck is hard to detect and I still think it's a good way (not everybody check the logs during streaming, if there is no log aggregation only at the end of the streaming job it will be collected) but we can add `failOnQueueThreshold=false` or something similar and declaring this as a soft threshold and printing out a warning if reached. `failOnQueueThreshold=true` would mean stop the query. If the user sees the warning and the query recovers properly the number can still be increased.

Reply:
Ok, I understand the argument, but I don't feel qualified to make a decision about whether this change in semantics is ok or not. Perhaps you can ping Jose enough until he or somebody else responds.

Try that again, and if that doesn't work, then I'll follow what you think is better here.

Reply:
Oh, sorry, I've just been following along here. I agree that a queue of 10k epochs stacked up indicates the query is never going to make progress. The tasks of a continuous processing query are all co-scheduled - if one or more of the tasks ends up that far behind, it's almost surely because it's unrecoverably broken.

Reply:
@vanzin I understand your point and introducing possible breaks is a legal fear. Thanks for offering such resolution.
@jose-torres thanks for jumping in and confirming it, extremely helpful!